src/App.js

```diff
@@ -1,21 +1,57 @@
-import React, { Component } from "react";
+import React from "react";
 import { Route, BrowserRouter, Switch } from "react-router-dom";
 import DashboardPage from "./pages/DashboardPage";
 import SignIn from "./modules/sign/components/Signin";
+import { createContext } from "react"
+import { useReducer} from "react"
+import { SetGitHubAuth, setGitLabAuth } from "./utils/misc.utils";
+
+const reducer = (state, { type, payload }) => {
+  switch (type) {
+    case "LOGIN":
+      return { ...state, loading: true };
+
+    case "LOGIN_SUCESS":
+      return { ...state, user: payload, loading: false };
+
+    case "LOGIN_FAILED":
+      return { ...state, user: {}, error: payload, loading: false };
+
+    case "GITLAB_AUTH":
+      return { ...state, isGitlabAuth: setGitLabAuth() }
+    
+    case "CLEAR_GITLAB_AUTH":
+      return { ...state, isGitlabAuth: SetGitHubAuth() }
```

Comment:
Again as I mentioned earlier, reducers should only return new state. But in this case, the SetGitHubAuth the side effect of setting localStorage data.